test/cql-pytest/test_in_operator.py

```diff
@@ -0,0 +1,80 @@
+# (at your option) any later version.
+#
+# Scylla is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+
+import pytest
+from cassandra import InvalidRequest
+from cassandra.cluster import Session, NoHostAvailable
+from cassandra.protocol import SyntaxException
+
+from util import unique_name
+
+
+@pytest.fixture(scope="module")
+def table_name(request, cql: Session, test_keyspace: str) -> str:
+    table = f"{test_keyspace}.{unique_name()}"
+
+    cql.execute(f"CREATE TABLE {table} (p int PRIMARY KEY, a ascii, b boolean, d double,"
+                f"mai map<ascii, int>, tup frozen<tuple<text, int>>, l list<text>)")
+    request.addfinalizer(lambda: cql.execute("DROP TABLE " + table))
+    return table
+
+
+@pytest.mark.parametrize("column,column_type", [
+    ("mai", "map<ascii, int>"),
+    ("l", "list<text>"),
+])
+def test_in_query_with_invalid_column_type(cql: Session, table_name: str, column: str, column_type: str):
+    expected_error = 'Error from server: code=2200 [Invalid query] message="' \
+                     f'Collection column \'{column}\' ({column_type}) cannot be restricted by a \'IN\' relation"'
+    with pytest.raises(InvalidRequest) as error:
+        cql.prepare(f"SELECT * FROM {table_name} where {column} IN (?, ?) ALLOW FILTERING;")
+    assert error.value.args[0] == expected_error
+
+
+def test_in_query_without_allow_filtering(cql: Session, table_name: str):
+    expected_error = 'Error from server: code=2200 [Invalid query] message="' \
+                     f'Cannot execute this query as it might involve data filtering and thus may have unpredictable' \
+                     f' performance. If you want to execute this query despite the performance unpredictability,' \
+                     f' use ALLOW FILTERING"'
+    with pytest.raises(InvalidRequest) as error:
+        cql.prepare(f"SELECT * FROM {table_name} where p IN (?, ?) AND a IN (?, ?);")
+    assert error.value.args[0] == expected_error
+
+
+def test_in_query_with_valid_product_limit(cql: Session, table_name: str):
+    limit = 100
+    query = f"SELECT * FROM {table_name} WHERE p IN {tuple(idx for idx in range(limit))} ALLOW FILTERING;"
+    cql.execute(query)
+
+
+def test_in_query_with_invalid_product_limit(cql: Session, table_name: str):
+    limit = 101
+    query = f"SELECT * FROM {table_name} WHERE p IN {tuple(idx for idx in range(limit))} ALLOW FILTERING;"
+    expected_error = f'Error from server: code=0000 [Server error] message="partition key cartesian product' \
+                     f' size {limit} is greater than maximum 100"'
+    with pytest.raises(expected_exception=NoHostAvailable) as error:
```

Comment:
I wonder if getting such an ugly internal server error is what we want to see here :-( InvalidRequest would have been much better. Of course, this is a Scylla problem, not a test bug, but it's good to know that we have this problem.

Reply:
Done